Fix equations sign in Beam module

nopria opened the pull request:
#### References to other Issues or PRs

Fixes #20204
See also #18045

#### Brief description of what is fixed or changed

Fix the sign of 2 equations used in the code:

1)  q =dV/dx   becomes   -q =dV/dx   (where q is distributed vertical load, V is shear, x is beam axis)
2)  EJ v'' = M   becomes   -EJ v'' = M   (where v is vertical displacement, M is moment, E is elastic modulus, J is moment of inertia)

The first equation is responsible (before fix) for wrong shear sign.
The second is responsible (before fix) for wrong results when used together with another equation present in code, **V = dM/dx**. In fact    **V = dM/dx**   and    **EJ v'' = M**   are incompatible, one of the two must have its sign changed, and changing sign to the latter equation allows to use the sign conventions most commonly used for internal forces ("discordant" shear and moment).

#### Other comments

- Optional TO DO: add an option to allow user to choose between "discordant" and "concordant" shear and moment convention.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* physics.continuum_mechanics
  * make beam module compute correctly internal forces
<!-- END RELEASE NOTES -->

Commit:
Fix equations sign

Fix the sign of 2 equations used in the code:

1)  q =dV/dx   becomes   -q =dV/dx   (where q is distributed vertical load, V is shear, x is beam axis)
2)  EI v'' = M   becomes   -EI v'' = M   (where v is vertical displacement, M is moment, E is elastic modulus, I is moment of inertia)

The first equation was responsible (before fix) for wrong shear sign.
The second was responsible (before fix) for wrong results when used together with the other equation V = dM/dx. In fact    V = dM/dx   and    EI v'' = M   are incompatible, one of the two must have its sign changed, and changing sign to the latter equation allows to use the sign conventions most commonly used for internal forces (discordant shear and moment).

TO DO: check and update documentation examples not used for testing.
Optional TO DO: add an option to allow user to choose between "discordant" and "concordant" shear and moment convention.

Fix sympy#20204
See also sympy#18045

Commit:
Update test_beam.py

Update tests to make equations with fixed sign pass them.

Commit:
Update description of allowed sign conventions

Commit:
Update documentation to reflect fix of equations sign

Commit:
Update documentation to reflect fix of equations sign

Commit:
More documentation update

Commit:
More test updates

Commit:
Yet more documentation updates

Commit:
Yet more test unit updates

Commit:
Yet 2 more test_beam.py update

Commit:
Restore unused expand line 160

Commit:
Some more test_beam.py update

nopria commented:
This commit fix a critical issue in Beam module, any idea on when it will be merged?

nopria commented:
The horrible bug explained in issue #20204, which is solved by this pull request, still affects new Sympy version just published. It's really a pity that this commit has not been merged yet.

oscarbenjamin commented:
I'm closing and reopening this to rerun the tests.

oscarbenjamin commented:
@moorepants can you review this?

moorepants commented:
@Psycho-Pirate looked at this in much more detail that I did. @Psycho-Pirate can you review?

Psycho-Pirate commented:
> @Psycho-Pirate looked at this in much more detail that I did. @Psycho-Pirate can you review?

Sure, I will look into it.

Psycho-Pirate commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -29,7 +29,10 @@ class Beam:
        While solving a beam bending problem, a user should choose its
        own sign convention and should stick to it. The results will
-       automatically follow the chosen sign convention.
+       automatically follow the chosen sign convention. However, the
+       chosen sign convention must respect the rule that on the positive
+       axis side of the section, a force giving positive shear yields a
+       negative moment.
```
I think this note might not be sufficient enough and beginner friendly in explaining what sign convention the user has to choose. Maybe adding an example might help. Something like this:
```
 For Example:   
    Given a right handed coordinate system with XYZ coordinates,
    the beam's length is assumed to be along the positive X axis.
    The loading forces are directed either in the positive or
    negative Y direction and the loading moments are about
    the positive or negative Z direction.
    The only thing that should be considered while choosing
    sign convention is that forces acting along positive
    Y direction(upward in this case) and moments acting along
    negative z direction(clockwise in this case) must have the same signs.
```
What are your thoughts?

Psycho-Pirate replied:
I have seen a lot of new users make this kind of mistake and getting wrong results with the code.

nopria replied:
Describing in a few words the allowed sign conventions it's not easy, maybe it's better to add an image like the following one. The curved arrow show the positive rotation. In orange is shown a possible (among many) alternate name for axes.

![github-allowed-signes-convention-opt](https://user-images.githubusercontent.com/4848618/102800319-fe532180-43b3-11eb-87d3-5de705a7d475.png)

Psycho-Pirate replied:
Yes that would be better. Go ahead!

nopria replied:
I've updated the note about allowed sign conventions and added a description image, but I don't know how to check the result through Sphinx (I guess).

Psycho-Pirate replied:
[See this.](https://docs.sympy.org/latest/documentation-style-guide.html?highlight=make%20html#build-the-documentation)
Please build the documentation and open the html file in a browser to see if the image is inserted or not. 
Also add a screenshot here of the webpage showing the image with the note.

nopria replied:
Running `make html` I get an error:

```
rm -rf _build/logo
mkdir -p _build/logo
python ./generate_logos.py -d
INFO:root: File saved: C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-notail.svg
INFO:root: File saved: C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-notailtext.svg
INFO:root: File saved: C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-notext.svg
ERROR:root:Return code is not 0: Command: rsvg-convert C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\src/logo\sympy.svg -f png -o C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-160px.png -h 160 -w 160
ERROR:root:return code: 1
mingw32-make: *** [Makefile:144: logo] Error 1
```

Psycho-Pirate replied:
There must be some unfulfilled dependencies. Let me build it on my system.

Psycho-Pirate replied:
I tried building it. The image is present in the wrong directory. Please move it to `sympy/doc/src/modules/physics/continuum_mechanics/` and delete the one in the current directory.

nopria replied:
I applied your commit suggestion and moved the image file to `sympy/doc/src/modules/physics/continuum_mechanics/`, but running `make html` I still get an error:

```
(nopria-sympy) C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc>make html
rm -rf _build/logo
mkdir -p _build/logo
python ./generate_logos.py -d
INFO:root: File saved: C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-notail.svg
INFO:root: File saved: C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-notailtext.svg
INFO:root: File saved: C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-notext.svg
ERROR:root:Return code is not 0: Command: rsvg-convert C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\src/logo\sympy.svg -f png -o C:\DATI\ingegneria\code\sympy-nopria\sympy-patch-1\doc\_build/logo\sympy-160px.png -h 160 -w 160
ERROR:root:return code: 1
mingw32-make: *** [Makefile:144: logo] Error 1
```

Psycho-Pirate replied:
![Screenshot from 2020-12-24 00-29-20](https://user-images.githubusercontent.com/50932406/103029425-cc8eb600-457f-11eb-91bc-17a693b96dd3.png)
Its working now.

Psycho-Pirate replied:
> I applied your [commit suggestion](https://github.com/sympy/sympy/pull/20431#discussion_r548012444) and moved the image file to `sympy/doc/src/modules/physics/continuum_mechanics/`, but running `make html` I still get an error:

I think you havent installed the dependencies required for sphinx properly. Have you installed librsvg?

nopria replied:
No, I didn't. After installing librsvg I still get other errors, but I don't care as long as it worked for you. Glad to know `beam.py` will finally give correct results. Next I'd like to improve `apply_support` which currently isn't well designed. Have a look at issue #20404 if you have time.

Psycho-Pirate replied:
> No, I didn't. After installing librsvg I still get other errors, but I don't care as long as it worked for you. 

When i first built the documentation i had to install librsvg and imagemagick along with sphinx, as far as i remember. My terminal did tell me what dependencies i was missing. 
>Next I'd like to improve `apply_support` which currently isn't well designed. Have a look at issue #20404 if you have time.

Sure, I'll look into it!

Psycho-Pirate reviewed (changes requested):
This looks great! I have added a few comments that must be addressed before merging.

Commit:
Update sign convention note

Adding text and a description image

Commit:
Add image describing allowed sign conventions

Psycho-Pirate commented on doc/src/modules/physics/continuum_mechanics/beam_problems.rst:
```diff
@@ -263,45 +263,45 @@ deflection is restricted at both the supports.
+>>> b.apply_load(8, 0, -1)
 >>> b.apply_load(R1, 10, -1)
 >>> b.apply_load(R2, 30, -1)
 >>> b.apply_load(120, 30, -2)
```
Please fix this as well. It would be great if we could fix all the doc errors in one pull request.

nopria replied:
Done, it was a typo in the problem description, now the description agrees with figure too.

Commit:
Fix description of example 2

Psycho-Pirate commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -29,7 +29,12 @@ class Beam:
+       side of beam's axis (in respect to current section), a loading force
+       giving positive shear yields a negative moment, as below (the
+       curved arrow shows the positive moment and rotation):
+       image:: allowed-sign-conventions.png
```
```suggestion

    .. image:: allowed-sign-conventions.png
```

Commit:
Fix note of example 2 description

Co-authored-by: Prakhar Saxena <50932406+Psycho-Pirate@users.noreply.github.com>

Commit:
Add image file

Used by note in example 2 of sympy/physics/continuum_mechanics/beam.py

Commit:
Delete allowed-sign-conventions.png

The image has been placed in the correct directory:  sympy/doc/src/modules/physics/continuum_mechanics/

Psycho-Pirate reviewed (approved):
Looks good to me! Great job!

Psycho-Pirate commented:
@moorepants Please get this merged as i do not have write access.

nopria commented:
> @moorepants Please get this merged as i do not have write access.

Any idea on why it hasn't been merged yet?

moorepants commented:
@nopria, please be patient, we are all volunteers.

Thanks for this addition. Based on @Psycho-Pirate's review, merging.
